google-cloud-storage/src/test/resources/META-INF/native-image/com/google/cloud/storage/native-image.properties

```diff
@@ -0,0 +1,26 @@
+Args = \
+ --initialize-at-build-time=com.google.cloud.conformance.storage.v1,\
+  com.google.protobuf,\
+  com.google.auth.oauth2,\
+  com.google.cloud.storage.conformance.retry,\
+  com.google.common.base.Charsets,\
+  com.google.gson.stream.JsonReader,\
+  com.google.api.client.util,\
+  com.google.api.client.http.javanet.NetHttpTransport,\
+  com.google.api.client.http.HttpTransport,\
+  com.google.api.client.json,\
+  com.google.common.io.BaseEncoding,\
+  com.google.common.math.IntMath$1,\
+  com.google.common.collect.Platform,\
+  com.google.gson.Gson,\
+  com.google.common.truth,\
+  com.google.common.collect,\
+  com.google.gson.internal.reflect,\
+  com.google.gson.internal.bind,\
+  com.google.gson.internal,\
+  com.google.gson.internal.sql.SqlTypesSupport,\
+  com.google.gson.FieldNamingPolicy$3,\
+  com.google.gson.LongSerializationPolicy$2
+
+
+
```

Comment:
Does this mean that `com.google.cloud.storage.conformance.retry.ITRetryConformanceTest#testCases` is being invoked during the building of the native image?

We're using `@RunWith(ParallelParameterized.class)` to resolve test cases relative to a classpath resource, and run them in parallel using a threadpool.

Reply:
While standard [Java initializes classes the first time they are accessed at run-time](https://docs.oracle.com/en/graalvm/enterprise/20/docs/reference-manual/native-image/ClassInitialization/), Native image compilation tries to do some optimizations were it initializes some classes at image build time to decrease the start-up time of an application. 

In our case, the `JunitFeature` which is brought in by the native-maven-plugin (plugin we use to run tests with native image compilation) explicitly initializes `Parameterized` at image build time. This causes `ParallelParameterized` and subsequently `com.google.cloud.storage.conformance.retry.ITRetryConformanceTest` and other classes `ITRetryConformanceTest` references to also be initialized at build time. While the class itself is initialized at build time, it still runs at run-time (after the native image has been built).

This [article](https://medium.com/graalvm/updates-on-class-initialization-in-graalvm-native-image-generation-c61faca461f7) has been helping me understand class initialization a bit in graal.